pkg/services/ngalert/services/alertmanager_config.go

```diff
@@ -0,0 +1,229 @@
+package services
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/notifier"
+	"github.com/grafana/grafana/pkg/services/ngalert/store"
+	"github.com/grafana/grafana/pkg/services/secrets"
+)
+
+type UnknownReceiverError struct {
+	UID string
+}
+
+func (e UnknownReceiverError) Error() string {
+	return fmt.Sprintf("unknown receiver: %s", e.UID)
+}
+
+// AlertmanagerConfigService is a domain-layer service which manages configs for the Grafana alertmanager.
+type AlertmanagerConfigService struct {
+	mam     *notifier.MultiOrgAlertmanager
+	secrets secrets.Service
+	store   AlertingStore
+	log     log.Logger
+}
+
+type AlertingStore interface {
+	GetLatestAlertmanagerConfiguration(ctx context.Context, query *models.GetLatestAlertmanagerConfigurationQuery) error
+}
+
+type Alertmanager interface {
+	// Configuration
+	SaveAndApplyConfig(ctx context.Context, config *definitions.PostableUserConfig) error
+	SaveAndApplyDefaultConfig(ctx context.Context) error
+	GetStatus() definitions.GettableStatus
+
+	// Silences
+	CreateSilence(ps *definitions.PostableSilence) (string, error)
+	DeleteSilence(silenceID string) error
+	GetSilence(silenceID string) (definitions.GettableSilence, error)
+	ListSilences(filter []string) (definitions.GettableSilences, error)
+
+	// Alerts
+	GetAlerts(active, silenced, inhibited bool, filter []string, receiver string) (definitions.GettableAlerts, error)
+	GetAlertGroups(active, silenced, inhibited bool, filter []string, receiver string) (definitions.AlertGroups, error)
+
+	// Testing
+	TestReceivers(ctx context.Context, c definitions.TestReceiversConfigBodyParams) (*notifier.TestReceiversResult, error)
+}
+
+func NewAlertmanagerConfigService(mam *notifier.MultiOrgAlertmanager, secrets secrets.Service, store AlertingStore, log log.Logger) *AlertmanagerConfigService {
+	return &AlertmanagerConfigService{
+		mam:     mam,
+		secrets: secrets,
+		store:   store,
+		log:     log,
+	}
+}
+
+func (s *AlertmanagerConfigService) GetAlertmanagerConfiguration(ctx context.Context, org int64) (definitions.GettableUserConfig, error) {
```

Comment:
This and below is factored out of existing handlers.

Logic is kept pretty much intact. Main changes include error handling (`fmt.Errorf`'s instead of `ErrResp`'s) and removal of mapping to status codes, which stayed in the layer above.